New Binding for Bose SoundTouch 10

xianlukas opened the pull request:
First version of a binding for the Bose SoundTouch 10 Speaker using webservices offered by speaker.
Please review it.

kaikreuzer commented:
Thanks! A very first question upfront: Wouldn't it be better to simply call the binding "bosesoundtouch"? I would assume that there are/will be other Soundtouch products in the portfolio that will use the same API, so SoundTouch 10 is only one specific thing type (even if it is the only supported one for the start).

xianlukas commented:
Hi, yes, this is correct, but I only had the chance to test on this specific devices. Other devices may have more and different functionalities. I also wrote this in the readme. I do not want to confuse people that all devices are fully supported.

kaikreuzer commented:
Understood, nonetheless, other people with other devices might be happy to test and enhance the binding once it is available. So I think it would be useful in the long run, to have the name changed. With some automatic refactoring support, it shouldn't be too hard to change, I think.
No need to do it right now, you can wait for a detailed review.

xianlukas commented:
Ok, I see the point. I refactor it after the detailed review. Thanks a lot for your fast feedback.

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ClassPath: .
+Import-Package: 
+ javax.jmdns;version="3.5.1",
```
Do you need those specific versions of the jmdns and jetty libraries? If not, please remove the version restrictions.

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Bundle-Name: BoseSoundTouch10 Binding
+Bundle-SymbolicName: org.openhab.binding.bosesoundtouch10;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
```
Please change version to 2.1.0

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/pom.xml:
```diff
@@ -0,0 +1,19 @@
+  <parent>
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>2.0.0-SNAPSHOT</version>
```
Please change version to 2.1.0

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/pom.xml:
```diff
@@ -0,0 +1,19 @@
+
+  <groupId>org.openhab.binding</groupId>
+  <artifactId>org.openhab.binding.bosesoundtouch10</artifactId>
+  <version>2.0.0-SNAPSHOT</version>
```
Please change version to 2.1.0

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch10/handler/BoseSoundTouch10Handler.java:
```diff
@@ -0,0 +1,378 @@
+                            break;
+                        }
+                        switch (source) {
+                            case "BLUETOOTH":
```
Please put constants somewhere for those fixed Strings. "BLUETOOTH", "AUX", "INTERNET_RADIO" etc.

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch10/handler/BoseSoundTouch10Handler.java:
```diff
@@ -0,0 +1,378 @@
+            client = new WebSocketClient();
+            try {
+                client.start();
+                URI echoUri = new URI(soundTouchURL.replace("http", "ws").replace("8090", "8080"));
```
Are the ports 8080 and 8090 fixed?

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch10/helper/WSHelper.java:
```diff
@@ -0,0 +1,160 @@
+        } catch (ProtocolException e) {
+            logger.debug("Error in GET protocol");
+        }
+        connection.setRequestProperty("Content-type", "text/xml");
```
I believe the header name should be "Content-Type" with an upper case "T".

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch10/handler/BoseSoundTouch10Handler.java:
```diff
@@ -0,0 +1,378 @@
+                    if (response.equals("error")) {
+                        break;
+                    }
+                    String volume = xmlHelper.parsePath("/volume/actualvolume", response);
```
The XmlHelper class already has a few Xpath constants. Could you put this in there as well?

hakan42 commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch10/helper/WSHelper.java:
```diff
@@ -0,0 +1,160 @@
+                response.append(inputLine);
+            }
+        } catch (IOException e) {
+            logger.debug("IOException on sending message to WS: " + message);
```
Please use parametrized logging instead of string concatenation.

kubawolanin commented on addons/binding/org.openhab.binding.bosesoundtouch10/about.html:
```diff
@@ -0,0 +1,36 @@
+<body lang="EN-US">
+<h2>About This Content</h2>
+ 
+<p>June 5, 2006</p>	
```
Please change [date to the current](http://docs.openhab.org/developers/development/guidelines.html#b-osgi-bundles).

xianlukas commented:
Hi @all, thanks for the reviews. I will improve the binding as soon as possible.

rtvb commented:
See Community

https://community.openhab.org/t/bose-soundtouch-binding/5678/72

tratho commented:
hi @all especially @xianlukas 
It seems we have some "problems". There is already a binding for bosesoundtouch under develoment. See the link @rtvb has posted.
@xianlukas it would be nice if you contact me (@tiknx) or @marvkis in the community forum

xianlukas commented:
hi,
@rtvb 
thanks for the hint, I will check that
@tratho 
sure I will, did not know about this development

xianlukas commented:
Hi @all,
thanks again for the reviews. Finally the improvements are available now.
version number updated or deleted where not needed
added contants for strings like INTERNET_RADIO
Ports are fixed, but changed to constants
using parametrized logging instead of string concatenation
changed Content-type to Content-Type in WSHelper.java
changed date in about.html to current

xianlukas commented:
Has anyone an idea why the build failed?

farrrell504 commented:
Are either @xianlukas or @marvkis / @tiknx working on this anymore?

xianlukas commented:
Yes, the mentions from review are done. The build error was something different which had nothing to do with this binding. So I am waiting to commit.

farrrell504 commented:
@xianlukas any idea when you will be committing this?

xianlukas commented:
@farrrell504 It is no quite clear for me what the next step is in this process.
Need to ask @kaikreuzer. From my point of view, everything is done, I just need @kaikreuzer to commit with him.

kaikreuzer commented:
Everyone is invited to review this against the guidelines.
@xianlukas In case you are feeling bored meanwhile, you might want to already do the refactoring to remove the "10" from the name. Besides this, I can only ask you for your patience!

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/OSGI-INF/BoseSoundTouchHandlerFactory.xml:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2014-2016 by the respective copyright holders.
```
It seems that this file has a missing  or invalid copyright header, can you fix all of them by running 

`mvn license:format` 

Please commit only the ones for your binding.

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/ESH-INF/i18n/bosesoundtouch10_xx_XX.properties:
```diff
@@ -0,0 +1,11 @@
+# binding
```
Please remove this unused file

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Bundle-SymbolicName: org.openhab.binding.bosesoundtouch;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.1.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
```
Can you change this to 1.8

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/README.md:
```diff
@@ -0,0 +1,44 @@
+| control | Player | Player control; currently only supports play/pause |
+
+## Full Example
+
```
The full example looks empty?

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/about.html:
```diff
@@ -0,0 +1,36 @@
+		There are <a href="http://products.bose.com/api-developer/assets/Bose_API_License_121114.pdf">Terms and Conditions</a> of use. 
+		</em></p>
+</body>
+</html>
```
Missing newline at the end of the file.

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/build.properties:
```diff
@@ -0,0 +1,6 @@
+bin.includes = META-INF/,\
+               .,\
+               OSGI-INF/,\
+               ESH-INF/
```
Missing newline and please add the about.html here.

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/BoseSoundTouchBindingConstants.java:
```diff
@@ -0,0 +1,34 @@
+    public static final String BINDING_ID = "bosesoundtouch";
+
+    // List of all Thing Type UIDs
+    public final static ThingTypeUID THING_TYPE_SOUNDTOUCH = new ThingTypeUID(BINDING_ID, "bosesoundtouch");
```
Non-preferred order of modifiers (It was like this in a template, which has been changed recently) correct order is `public static final` See also: http://stackoverflow.com/questions/16731240/what-is-a-reasonable-order-of-java-modifiers-abstract-final-public-static-e#16732059

martinvw replied:
the name of thing is now: `bosesoundtouch:bosesoundtouch`

is there a better solution? `bosesoundtouch:soundtouch10` maybe?

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/handler/BoseSoundTouchHandler.java:
```diff
@@ -0,0 +1,397 @@
+ * The {@link BoseSoundTouchHandler} is responsible for handling commands, which are
+ * sent to one of the channels, and handling messages coming from the speaker.
+ *
+ * @author syracom - Initial contribution
```
Please use a real name here, also check other occurrences

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/handler/BoseSoundTouchHandler.java:
```diff
@@ -0,0 +1,397 @@
+                        default:
+                            break;
+                    }
+                }
```
Missing break for this switch-case?

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/handler/BoseSoundTouchHandler.java:
```diff
@@ -0,0 +1,397 @@
+                    }
+                }
+            case BoseSoundTouchBindingConstants.CHANNEL_SOURCE:
+                if (command instanceof RefreshType) {
```
This body is too big to single case of a switch, please make it a dedicated method.

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/handler/BoseSoundTouchHandler.java:
```diff
@@ -0,0 +1,397 @@
+                if (client != null) {
+                    client.stop();
+                }
+            } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/handler/BoseSoundTouchHandler.java:
```diff
@@ -0,0 +1,397 @@
+                request.setSubProtocols("gabbo");
+                client.connect(socketListener, echoUri, request);
+                return true;
+            } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/helper/WSHelper.java:
```diff
@@ -0,0 +1,160 @@
+        String url = deviceUrl + key;
+
+        BufferedReader in = null;
+        StringBuffer response = new StringBuffer();
```
`StringBuilder` is faster than `StringBuffer` because it's not synchronized. So you should always use `StringBuilder`
 
And the thread safety which is delivered by `StringBuffer` is usually not enough because different append's can still be interleaved which still ruin your `StringBuffer`.

Just memorize to use the one which is the furthest in the Alphabet.

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/helper/WSHelper.java:
```diff
@@ -0,0 +1,160 @@
+        try {
+            connection.setRequestMethod("POST");
+        } catch (ProtocolException e) {
+            logger.debug("Error in GET protocol");
```
Error in get or post protocol?

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/helper/WSHelper.java:
```diff
@@ -0,0 +1,160 @@
+        try {
+            con.setRequestMethod("GET");
+        } catch (ProtocolException e) {
+            logger.debug("Error in GET protocol");
```
Can you continue normally and does this have no real impact?

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/helper/XmlHelper.java:
```diff
@@ -0,0 +1,192 @@
+            XPath xPath = XPathFactory.newInstance().newXPath();
+            result = xPath.compile(expression).evaluate(xmlDocument);
+        } catch (XPathExpressionException e) {
+            logger.debug("Exception on XMLHelper");
```
Do you really want to 'ignore' the exception like this, by debug logging and just continuing with the output ""?

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/src/main/java/org/openhab/binding/bosesoundtouch/internal/discovery/BoseSoundTouchDiscoveryParticipant.java:
```diff
@@ -0,0 +1,66 @@
+            return null;
+        }
+    }
+}
```
Missing newline at the end of the file.

martinvw commented on addons/binding/org.openhab.binding.bosesoundtouch10/pom.xml:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
The project is missing from the parent pom.xml 

To have the binding being picked up by the distro, you furthermore need to add it to the feature.xml, again at the alphabetically correct position. If you have a dependency on some transport bundle (e.g. upnp, mdns or serial), make sure to add a line for this dependency as well (see the other bindings as an example).

xianlukas commented:
Hi @martinvw,
thanks a lot for the review. I will take care of that as soon as possible.

kaikreuzer commented:
Before putting further effort into it, please first let us come to a conclusion regarding https://github.com/openhab/openhab2-addons/pull/2233#issuecomment-299304925.

kaikreuzer commented:
> Before putting further effort into it, please first let us come to a conclusion regarding #2233 (comment).

From what I see from the community, people are choosing the implementation from #2233 as it has more features implemented. Without wanting to judge any PR to be better than another, I would suggest to follow the community and thus close this PR. I hope this is fine for you @xianlukas - I can understand if it is frustrating to see a PR not ending up in the distro, but I hope for your understanding. Thanks for your work nonetheless!

xianlukas commented:
@kaikreuzer 
fine for me
